x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/role/BulkPutRoleRequestBuilderFactory.java

```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.security.action.role;
+
+import org.elasticsearch.client.internal.Client;
+
+public interface BulkPutRoleRequestBuilderFactory {
+    BulkPutRoleRequestBuilder create(Client client, boolean restrictRequest);
+
+    class Default implements BulkPutRoleRequestBuilderFactory {
+        @Override
+        public BulkPutRoleRequestBuilder create(Client client, boolean restrictRequest) {
+            // This needs to be added when Bulk API is made public in serverless
+            return new BulkPutRoleRequestBuilder(client);
+        }
+    }
+}
```

Comment:
This is needed to allow us to inject another builder through SPI when running in serverless to enforce custom roles validation for the bulk api.

Reply:
My suggestion is to make this PR simpler and not think about validation for serverless roles. Your call.